### PR TITLE
feat: Test maxSlippage with ERC4626 donation attack

### PR DESCRIPTION
**Pull request by hacker-DOM**
- **Backup: Before refactoring to use interface instead of address.**
- **Backup: Passing tests, setup complete.**
- **Backup: Right before removing morpho vaults to test donation attack.**
- **Backup: MetaMorpho v1.1 full attack [doesn't work].**
- **S Feat: Small changes.**
- **S Feat: Finalize morpho vault attack.**

## Summary by CodeRabbit

* **Tests**
  * Added two mainnet-fork end-to-end tests for an ERC4626 donation-attack scenario: one verifying slippage-protected failure and one verifying successful deposit after an attack.
  * Exercises realistic governance/timelock and vault setup, per-function rate limits, relayer interactions, and attacker funding flow.
  * Validates post-attack vault state: shares, assets, total supply, asset↔share conversions and rounding edge cases.

### CODE REVIEW

File: test/mainnet-fork/ERC4626DonationAttack.t.sol [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import { IMetaMorpho, Id } from "metamorpho/interfaces/IMetaMorpho.sol";
+
+import { MarketParamsLib }               from "morpho-blue/src/libraries/MarketParamsLib.sol";
+import { IMorpho, MarketParams, Market } from "morpho-blue/src/interfaces/IMorpho.sol";
+
+import "./ForkTestBase.t.sol";
+
+contract ERC4626DonationAttackTestBase is ForkTestBase {
+
+    IMetaMorpho morpho_vault = IMetaMorpho(0xe41a0583334f0dc4E023Acd0bFef3667F6FE0597);
```
Comment (lucas-manuel) [inline test/mainnet-fork/ERC4626DonationAttack.t.sol]: camelCase for all values in this PR

File: test/mainnet-fork/ERC4626DonationAttack.t.sol [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        bytes32 depositKey  = RateLimitHelpers.makeAddressKey(mainnetController.LIMIT_4626_DEPOSIT(),  address(morpho_vault));
+        bytes32 withdrawKey = RateLimitHelpers.makeAddressKey(mainnetController.LIMIT_4626_WITHDRAW(), address(morpho_vault));
+
+        // Basic validation
+        assertEq(keccak256(abi.encode(morpho_vault.symbol())), keccak256(abi.encode("sparkUSDS")));
+        assertEq(morpho_vault.totalAssets(),                   0);
+        assertEq(morpho_vault.totalSupply(),                   0);
+
+        // Initialization
+        vm.startPrank(Ethereum.SPARK_PROXY);
```
Comment (lucas-manuel) [inline test/mainnet-fork/ERC4626DonationAttack.t.sol:48]: New line after this and before stopPrank to make it clearer that its for the whole block between

File: test/mainnet-fork/ERC4626DonationAttack.t.sol [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        // Now that the market has a non-zero cap, set the supply queue order
+        Id[] memory supplyOrder = new Id[](1);
+        supplyOrder[0] = marketId;
+        morpho_vault.setSupplyQueue(supplyOrder);
+        vm.stopPrank();
+
+        assertEq(morpho_vault.curator(),      curator);
+        assertEq(morpho_vault.guardian(),     guardian);
+        assertEq(morpho_vault.feeRecipient(), fee_recipient);
+        assertTrue(morpho_vault.isAllocator(allocator));
```
Comment (lucas-manuel) [inline test/mainnet-fork/ERC4626DonationAttack.t.sol]: New line

File: test/mainnet-fork/ERC4626DonationAttack.t.sol [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    function _getBlock() internal override pure returns (uint256) {
+        return 22932160;  // July 16, 2025
+    }
+
+}
+
+contract ERC4626DonationAttack is ERC4626DonationAttackTestBase {
+
+    function test_donationAttackERC4626_usds() external {
+        address mallory = makeAddr("mallory");
```
Comment (lucas-manuel) [inline test/mainnet-fork/ERC4626DonationAttack.t.sol]: attacker = "attacker"

File: test/mainnet-fork/ERC4626DonationAttack.t.sol [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            marketParams, 1_000_000e18, 0, address(morpho_vault), hex""
+        );
+        vm.stopPrank();
+
+        assertEq(assets, 1_000_000e18);
+        assertEq(shares, uint256(1_000_000e18) * market.totalSupplyShares / market.totalSupplyAssets);
+        assertEq(shares, 1e30);
+
+        assertEq(morpho_vault.balanceOf(mallory), 1);
+        assertEq(morpho_vault.totalSupply(), 1);
```
Comment (lucas-manuel) [inline test/mainnet-fork/ERC4626DonationAttack.t.sol]: Align

File: test/mainnet-fork/ERC4626DonationAttack.t.sol [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        try mainnetController.depositERC4626(address(morpho_vault), 2_000_000e18) {
+            // The deposit went through. The only time this is permissible is if the attack had no
+            // effect.
+            uint256 assetsOfProxy = morpho_vault.convertToAssets(morpho_vault.balanceOf(address(almProxy)));
+            assertEq(assetsOfProxy,                             2_000_000e18);
+            assertEq(morpho_vault.balanceOf(address(almProxy)), 2_000_000e24);
+        } catch Error(string memory reason) {
+            // The deposit was correctly reverted.
+            assertEq(reason, "MainnetController/slippage-too-high");
+        }
```
Comment (lucas-manuel) [inline test/mainnet-fork/ERC4626DonationAttack.t.sol]: This should be two separate tests, one with a really conservative slippage set and one with a loose slippage set and show success and revert cases. Abstract attack setup into a helper function. Tabish did the same thing for the AAVE testing so can follow his implementation for this.

Should assert result of successful attack, showing lost funds.

File: test/mainnet-fork/ERC4626DonationAttack.t.sol (hunk 1 above)
Comment (hacker-DOM) [reply test/mainnet-fork/ERC4626DonationAttack.t.sol]: ✅

File: test/mainnet-fork/ERC4626DonationAttack.t.sol (hunk 2 above)
Comment (hacker-DOM) [reply test/mainnet-fork/ERC4626DonationAttack.t.sol:48]: ✅

File: test/mainnet-fork/ERC4626DonationAttack.t.sol (hunk 3 above)
Comment (hacker-DOM) [reply test/mainnet-fork/ERC4626DonationAttack.t.sol]: ✅

File: test/mainnet-fork/ERC4626DonationAttack.t.sol (hunk 4 above)
Comment (hacker-DOM) [reply test/mainnet-fork/ERC4626DonationAttack.t.sol]: ✅

File: test/mainnet-fork/ERC4626DonationAttack.t.sol (hunk 5 above)
Comment (hacker-DOM) [reply test/mainnet-fork/ERC4626DonationAttack.t.sol]: ✅

Comment (hacker-DOM) [reply test/mainnet-fork/ERC4626DonationAttack.t.sol]: ✅

File: test/mainnet-fork/ERC4626DonationAttack.t.sol (hunk 6 above)
Comment (hacker-DOM) [reply test/mainnet-fork/ERC4626DonationAttack.t.sol]: ✅

File: test/mainnet-fork/ERC4626DonationAttack.t.sol [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        assertEq(market.totalSupplyAssets, 36_095_481.319542091092211965e18); // ~36M USDS
+        assertEq(market.totalSupplyShares, 36_095_481.319542091092211965000000e24);
+
+        deal(address(usds), attacker, 1_000_000e18 + 1);
+
+        vm.startPrank(attacker);
+        usds.approve(address(morphoVault), 1);
+        morphoVault.deposit(1, attacker);
+        usds.approve(address(morpho), 1_000_000e18);
+        // Donation attack
```
Comment (lucas-manuel) [inline test/mainnet-fork/ERC4626DonationAttack.t.sol]: Donation attack performed by donating shares of morpho market supply to morpho vault

File: test/mainnet-fork/ERC4626DonationAttack.t.sol [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    function test_depositERC4626_donationAttackSuccess() external {
+        // Set max slippage to (close to) 100%
+        vm.prank(Ethereum.SPARK_PROXY);
+        mainnetController.setMaxSlippage(address(morphoVault), 1);
+
+        _doAttack();
+        vm.prank(relayer);
+        uint256 shares = mainnetController.depositERC4626(address(morphoVault), 2_000_000e18);
+
+        // We can compute:
```
Comment (lucas-manuel) [inline test/mainnet-fork/ERC4626DonationAttack.t.sol]: No "we" or "us" language

File: test/mainnet-fork/ERC4626DonationAttack.t.sol (hunk 7 above)
Comment (lucas-manuel) [reply test/mainnet-fork/ERC4626DonationAttack.t.sol]: Also new line above

File: test/mainnet-fork/ERC4626DonationAttack.t.sol (hunk 8 above)
Comment (hacker-DOM) [reply test/mainnet-fork/ERC4626DonationAttack.t.sol]: ✅

File: test/mainnet-fork/ERC4626DonationAttack.t.sol (hunk 7 above)
Comment (hacker-DOM) [reply test/mainnet-fork/ERC4626DonationAttack.t.sol]: ✅
